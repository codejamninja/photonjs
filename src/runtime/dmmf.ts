import { DMMF } from './dmmf-types'
import { Dictionary, keyBy } from './utils/common'

export class DMMFClass implements DMMF.Document {
  public datamodel: DMMF.Datamodel
  public schema: DMMF.Schema
  public mappings: DMMF.Mapping[]
  public queryType: DMMF.OutputType
  public mutationType: DMMF.OutputType
  public outputTypes: DMMF.OutputType[]
  public outputTypeMap: Dictionary<DMMF.OutputType> = {}
  public inputTypes: DMMF.InputType[]
  public inputTypeMap: Dictionary<DMMF.InputType>
  public enumMap: Dictionary<DMMF.Enum>
  public modelMap: Dictionary<DMMF.Model>
  constructor({ datamodel, schema, mappings }: DMMF.Document) {
    this.datamodel = datamodel
    this.schema = schema
    this.mappings = mappings
    this.enumMap = this.getEnumMap()
    this.queryType = this.getQueryType()
    this.mutationType = this.getMutationType()
    this.modelMap = this.getModelMap()
    this.outputTypes = this.getOutputTypes()

    this.resolveOutputTypes(this.outputTypes)

    this.inputTypes = this.schema.inputTypes
    this.resolveInputTypes(this.inputTypes)
    this.inputTypeMap = this.getInputTypeMap()
    this.resolveFieldArgumentTypes(this.outputTypes, this.inputTypeMap)

    this.outputTypeMap = this.getMergedOutputTypeMap()

    // needed as references are not kept
    this.queryType = this.outputTypeMap.Query
    this.mutationType = this.outputTypeMap.Mutation
    this.outputTypes = this.outputTypes
  }
  public getField(fieldName: string) {
    return (
      // TODO: create lookup table for Query and Mutation
      this.queryType.fields.find(f => f.name === fieldName) || this.mutationType.fields.find(f => f.name === fieldName)
    )
  }
  protected outputTypeToMergedOutputType = (outputType: DMMF.OutputType): DMMF.OutputType => {
    const model = this.modelMap[outputType.name]
    return {
      ...outputType,
      isEmbedded: model ? model.isEmbedded : false,
      fields: outputType.fields,
    }
  }
  protected resolveOutputTypes(types: DMMF.OutputType[]) {
    for (const typeA of types) {
      for (const fieldA of typeA.fields) {
        for (const typeB of types) {
          if (typeof fieldA.outputType.type === 'string') {
            if (fieldA.outputType.type === typeB.name) {
              fieldA.outputType.type = typeB
            } else if (this.enumMap[fieldA.outputType.type]) {
              fieldA.outputType.type = this.enumMap[fieldA.outputType.type]
            }
          }
        }
      }
    }
  }
  protected resolveInputTypes(types: DMMF.InputType[]) {
    for (const typeA of types) {
      for (const fieldA of typeA.fields) {
        for (const typeB of types) {
          fieldA.inputType.forEach((inputType, index) => {
            if (typeof inputType.type === 'string') {
              if (inputType.type === typeB.name) {
                fieldA.inputType[index].type = typeB
              } else if (this.enumMap[inputType.type]) {
                fieldA.inputType[index].type = this.enumMap[inputType.type]
              }
            }
          })
        }
      }
    }
  }
  protected resolveFieldArgumentTypes(types: DMMF.OutputType[], inputTypeMap: Dictionary<DMMF.InputType>) {
    for (const type of types) {
      for (const field of type.fields) {
        for (const arg of field.args) {
          arg.inputType.forEach((t, index) => {
            if (typeof t.type === 'string') {
              if (inputTypeMap[t.type]) {
                arg.inputType[index].type = inputTypeMap[t.type]
              } else if (this.enumMap[t.type]) {
                arg.inputType[index].type = this.enumMap[t.type]
              }
            }
          })
        }
      }
    }
  }
  protected getQueryType(): DMMF.OutputType {
    return this.schema.outputTypes.find(t => t.name === 'Query')!
  }
  protected getMutationType(): DMMF.OutputType {
    return this.schema.outputTypes.find(t => t.name === 'Mutation')!
  }
  protected getOutputTypes(): DMMF.OutputType[] {
    return this.schema.outputTypes.map(this.outputTypeToMergedOutputType)
  }
  protected getEnumMap(): Dictionary<DMMF.Enum> {
    return keyBy(this.schema.enums, e => e.name)
  }
  protected getModelMap(): Dictionary<DMMF.Model> {
    return keyBy(this.datamodel.models, m => m.name)
  }
  protected getMergedOutputTypeMap(): Dictionary<DMMF.OutputType> {
    return keyBy(this.outputTypes, t => t.name)
  }
  protected getInputTypeMap(): Dictionary<DMMF.InputType> {
    return keyBy(this.schema.inputTypes, t => t.name)
  }
}
