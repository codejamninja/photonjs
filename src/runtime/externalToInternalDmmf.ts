import { DMMF as ExternalDMMF } from '@prisma/generator-helper'
import pluralize from 'pluralize'
import { DMMF } from './dmmf-types'
import { lowerCase } from './utils/common'

function transformFieldKind(model: ExternalDMMF.Model): DMMF.Model {
  return {
    ...model,
    fields: model.fields.map(field => ({ ...field, kind: field.kind === 'relation' ? ('object' as any) : field.kind })),
  }
}

function transformDatamodel(datamodel: ExternalDMMF.Datamodel): DMMF.Datamodel {
  return {
    enums: datamodel.enums,
    models: datamodel.models.map(transformFieldKind),
  }
}

/**
 * Turns type: string into type: string[] for all args in order to support union input types
 * @param document
 */
export function externalToInternalDmmf(document: ExternalDMMF.Document): DMMF.Document {
  return {
    datamodel: transformDatamodel(document.datamodel),
    mappings: getMappings(document.mappings),
    schema: transformSchema(document.schema),
  }
}

function transformSchema(schema: ExternalDMMF.Schema): DMMF.Schema {
  return {
    enums: schema.enums,
    inputTypes: schema.inputTypes.map(t => ({ ...t, fields: transformArgs(t.fields) })),
    outputTypes: schema.outputTypes.map(o => ({
      ...o,
      fields: o.fields.map(f => ({ ...f, args: transformArgs(f.args) })),
    })),
  }
}

function transformArgs(args: ExternalDMMF.SchemaArg[]): DMMF.SchemaArg[] {
  return args.map(transformArg)
}

function fixOrderByEnum(arg: ExternalDMMF.SchemaArg): ExternalDMMF.SchemaArg {
  if (arg.name === 'orderBy' && arg.inputType.type.endsWith('OrderByInput')) {
    return {
      name: arg.name,
      inputType: {
        isList: arg.inputType.isList,
        isRequired: arg.inputType.isRequired,
        type: arg.inputType.type,
        kind: 'object',
      },
    }
  }
  return arg
}

function transformArg(argBefore: ExternalDMMF.SchemaArg): DMMF.SchemaArg {
  const arg = fixOrderByEnum(argBefore)
  return {
    name: arg.name,
    inputType: [arg.inputType],
  }
}

function getMappings(mappings: ExternalDMMF.Mapping[]): DMMF.Mapping[] {
  return mappings.map((mapping: any) => ({
    model: mapping.model,
    plural: pluralize(lowerCase(mapping.model)),
    findOne: mapping.findSingle || mapping.findOne,
    findMany: mapping.findMany,
    create: mapping.createOne || mapping.createSingle || mapping.create,
    delete: mapping.deleteOne || mapping.deleteSingle || mapping.delete,
    update: mapping.updateOne || mapping.updateSingle || mapping.update,
    deleteMany: mapping.deleteMany,
    updateMany: mapping.updateMany,
    upsert: mapping.upsertOne || mapping.upsertSingle || mapping.upsert,
    aggregate: mapping.aggregate,
  }))
}
