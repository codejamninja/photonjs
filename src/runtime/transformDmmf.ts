import { DMMF } from './dmmf-types'
import { Dictionary, uniqBy } from './utils/common'

export function transformDmmf(document: DMMF.Document): DMMF.Document {
  const doc = transformOrderInputTypes(transformWhereInputTypes(document))
  return {
    datamodel: doc.datamodel,
    mappings: doc.mappings,
    schema: {
      enums: doc.schema.enums,
      rootMutationType: doc.schema.rootMutationType,
      rootQueryType: doc.schema.rootQueryType,
      outputTypes: filterOutputTypes(doc.schema.outputTypes),
      inputTypes: makeWhereUniqueInputsRequired(filterInputTypes(doc.schema.inputTypes)),
    },
  }
}

function filterInputTypes(types: DMMF.InputType[]): DMMF.InputType[] {
  return uniqBy(types, o => o.name)
}

function filterOutputTypes(types: DMMF.OutputType[]): DMMF.OutputType[] {
  return uniqBy(types, o => o.name)
}

function transformOrderInputTypes(document: DMMF.Document): DMMF.Document {
  const inputTypes: DMMF.InputType[] = document.schema.inputTypes
  const enums: DMMF.Enum[] = [
    {
      name: 'OrderByArg',
      values: ['asc', 'desc'],
    },
  ]
  for (const type of document.schema.enums) {
    if (!type.name.endsWith('OrderByInput')) {
      enums.push(type)
      continue
    }
    const argNames = type.values.reduce<string[]>((acc, curr) => {
      if (curr.endsWith('ASC')) {
        const index = curr.lastIndexOf('_ASC')
        acc.push(curr.slice(0, index))
      }
      return acc
    }, [])
    const inputType: DMMF.InputType = {
      name: type.name,
      atLeastOne: true,
      atMostOne: true,
      isOrderType: true,
      fields: argNames.map(name => ({
        name,
        inputType: [
          {
            type: 'OrderByArg',
            isList: false,
            isRequired: false,
            kind: 'enum' as any,
          },
        ],
        isRelationFilter: false,
      })),
    }
    inputTypes.push(inputType)
  }

  return {
    datamodel: document.datamodel,
    mappings: document.mappings,
    schema: {
      ...document.schema,
      inputTypes,
      enums,
    },
  }
}

function makeWhereUniqueInputsRequired(inputTypes: DMMF.InputType[]): DMMF.InputType[] {
  return inputTypes.map(inputType => {
    if (inputType.name.endsWith('WhereUniqueInput')) {
      inputType.atLeastOne = true
    }
    return inputType
  })
}

function getFieldType(field: DMMF.Field): string {
  if (field.default && field.default.name === 'uuid') {
    return 'UUID'
  }
  return field.type
}

function transformWhereInputTypes(document: DMMF.Document): DMMF.Document {
  const types = document.schema.inputTypes
  const inputTypes: DMMF.InputType[] = []
  const filterTypes: Dictionary<DMMF.InputType> = {}
  for (const type of types) {
    if (!type.name.endsWith('WhereInput')) {
      inputTypes.push(type)
      continue
    }

    // lastIndexOf necessary if a type is called "WhereInput"
    let index = type.name.lastIndexOf('WhereInput')
    let modelName = type.name.slice(0, index)
    let model = document.datamodel.models.find(m => m.name === modelName)!
    if (!model) {
      index = type.name.lastIndexOf('ScalarWhereInput')
      modelName = type.name.slice(0, index)
      model = document.datamodel.models.find(m => m.name === modelName)!
    }
    if (!model) {
      inputTypes.push(type)
      continue
    }
    const whiteList = ['AND', 'OR', 'NOT']
    whiteList.push(...model.fields.filter(f => f.kind === 'object' && !f.isList).map(f => f.name))

    const fields = type.fields.filter(a => whiteList.includes(a.name)).map(a => ({ ...a, isRelationFilter: true }))

    const filterTypesList = model.fields
      // filter out scalar lists as Prisma doesn't have filters for them
      // also filter out object non-lists, as we don't need to transform them
      .filter(f => (f.kind === 'object' ? f.isList : !f.isList))
      .map(f => {
        if (!filterTypes[getFilterName(getFieldType(f), f.isRequired || f.kind === 'object')]) {
          filterTypes[getFilterName(getFieldType(f), f.isRequired || f.kind === 'object')] = makeFilterType(
            getFieldType(f),
            f.isRequired,
            f.kind !== 'object',
          )
        }

        const typeList: DMMF.SchemaArgInputType[] = []
        if (f.kind !== 'object') {
          typeList.push({
            isList: f.isList,
            isRequired: false,
            kind: f.kind,
            type: getFieldType(f),
          })
        }
        typeList.push({
          type: getFilterName(getFieldType(f), f.isRequired || f.kind === 'object'),
          isList: false,
          isRequired: false,
          kind: 'object',
        })

        // for optional scalars you can directly provide null
        if (!f.isRequired && f.kind !== 'object') {
          typeList.push({ type: 'null', isList: false, isRequired: false, kind: 'scalar' })
        }

        return {
          name: f.name,
          inputType: typeList,
          isRelationFilter: false,
        }
      })

    // NOTE: list scalar fields don't have where arguments!
    fields.unshift(...filterTypesList)
    const newType: DMMF.InputType = {
      name: type.name,
      fields,
      isWhereType: true,
      atLeastOne: false,
    }
    inputTypes.push(newType)
  }
  const scalarFilters = Object.values(filterTypes)
  inputTypes.push(...scalarFilters)

  return {
    datamodel: document.datamodel,
    mappings: document.mappings,
    schema: {
      ...document.schema,
      inputTypes,
    },
  }
}

function getFilterName(type: string, isRequired: boolean) {
  return `${isRequired ? '' : 'Nullable'}${type}Filter`
}

function getWhereInputName(type: string) {
  return `${type}WhereInput`
}

function makeFilterType(type: string, isRequired: boolean, isScalar: boolean): DMMF.InputType {
  return {
    name: getFilterName(type, isRequired || !isScalar),
    fields: isScalar ? getScalarFilterArgs(type, isRequired) : getRelationFilterArgs(type),
    atLeastOne: false,
  }
}

function getRelationFilterArgs(type: string): DMMF.SchemaArg[] {
  return getScalarArgs(['every', 'some', 'none'], [getWhereInputName(type)], undefined, 'object')
}

function getScalarFilterArgs(type: string, isRequired: boolean, isEnum = false): DMMF.SchemaArg[] {
  if (isEnum) {
    return [...getBaseFilters(type, isRequired), ...getInclusionFilters(type)]
  }
  switch (type) {
    case 'String':
    case 'ID':
    case 'UUID':
      return [
        ...getBaseFilters(type, isRequired),
        ...getInclusionFilters(type),
        ...getAlphanumericFilters(type),
        ...getStringFilters(type),
      ]
    case 'Int':
    case 'Float':
    case 'DateTime':
      return [...getBaseFilters(type, isRequired), ...getInclusionFilters(type), ...getAlphanumericFilters(type)]
    case 'Boolean':
      return [...getBaseFilters(type, isRequired)]
  }

  return []
}

function getBaseFilters(type: string, isRequired: boolean): DMMF.SchemaArg[] {
  const filterName = getFilterName(type, isRequired)
  // TODO: reintroduce AND, NOT, OR
  const nullArray = isRequired ? [] : ['null']
  return [
    ...getScalarArgs(['equals'], [type, ...nullArray]),
    ...getScalarArgs(['not'], [type, ...nullArray, filterName]),
  ]
}

function getStringFilters(type: string): DMMF.SchemaArg[] {
  return getScalarArgs(['contains', 'startsWith', 'endsWith'], [type])
}

function getAlphanumericFilters(type: string): DMMF.SchemaArg[] {
  return getScalarArgs(['lt', 'lte', 'gt', 'gte'], [type])
}

function getInclusionFilters(type: string): DMMF.SchemaArg[] {
  return getScalarArgs(['in', 'notIn'], [type], true)
}

function getScalarArgs(
  names: string[],
  type: string[],
  isList = false,
  kind: DMMF.FieldKind = 'scalar',
): DMMF.SchemaArg[] {
  return names.map(name => getScalarArg(name, type, isList, kind))
}

function getScalarArg(name: string, type: string[], isList, kind: DMMF.FieldKind = 'scalar'): DMMF.SchemaArg {
  return {
    name,
    isRelationFilter: kind === 'object',
    inputType: type.map(t => ({
      isList,
      isRequired: false,
      kind,
      type: t,
    })),
  }
}
