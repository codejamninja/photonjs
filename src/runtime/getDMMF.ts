import { DMMF } from '@prisma/generator-helper'
import { getDMMF as getRawDMMF, GetDMMFOptions } from '@prisma/sdk'
import { DMMF as PhotonDMMF } from './dmmf-types'
import { externalToInternalDmmf } from './externalToInternalDmmf'
import { transformDmmf } from './transformDmmf'

export function getPhotonDMMF(dmmf: DMMF.Document): PhotonDMMF.Document {
  return transformDmmf(externalToInternalDmmf(dmmf))
}

// Mostly used for tests
export async function getDMMF(
  options: GetDMMFOptions,
): Promise<PhotonDMMF.Document> {
  const dmmf = await getRawDMMF(options)
  return getPhotonDMMF(dmmf)
}
