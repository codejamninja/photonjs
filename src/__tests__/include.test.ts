import stripAnsi from 'strip-ansi'
import { blog } from '../fixtures/blog'
import { DMMFClass } from '../runtime/dmmf'
import { makeDocument } from '../runtime/query'
import { getDMMF } from '../runtime/getDMMF'

let dmmf
beforeAll(async () => {
  const dmmfDocument = await getDMMF({ datamodel: blog })
  dmmf = new DMMFClass(dmmfDocument)
})

describe('include validation', () => {
  test('deep include query', () => {
    const ast = {
      include: {
        author: {
          include: {
            posts: true,
          },
        },
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyPost',
    })

    expect(String(document)).toMatchSnapshot()
    expect(() => document.validate(ast)).not.toThrow()
  })

  test('dont allow empty include statements', () => {
    const ast = {
      include: {},
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyPost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast, false)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
        "
        Invalid \`photon.findManyPost()\` invocation:

        {
          include: {
        ?   author?: true
          }
        }


        The \`include\` statement for type Post must not be empty. Available options are listed in green.
        "
      `)
    }
  })

  // Why do we allow it with include but not with select?
  // Very simple, because with select a statement with only false properties is useless
  // Why do we throw for an empty object? Also very simple: We want to help people to explore
  // the API through errors!
  test('allow include statement with only false properties', () => {
    const ast = {
      include: {
        author: true,
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyPost',
    })

    expect(String(document)).toMatchSnapshot()
    expect(() => document.validate(ast)).not.toThrow()
  })

  test('allow deep include without another include', () => {
    const ast = {
      include: {
        posts: { first: 20 },
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyUser',
    })

    expect(String(document)).toMatchSnapshot()
    expect(() => document.validate(ast)).not.toThrow()
  })

  test('handle scalar fields special', () => {
    const ast = {
      include: {
        id: true,
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyPost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast, false)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
                "
                Invalid \`photon.id()\` invocation:

                {
                  include: {
                    id: true,
                    ~~
                ?   author?: true
                  }
                }


                Invalid scalar field \`id\` for include statement on model Post. Available options are listed in green.
                Note, that include statements only accept relation fields.
                "
            `)
    }
  })

  test('catch unknown field name', () => {
    const ast = {
      include: {
        mauthor: true,
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyPost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast, false)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
                        "
                        Invalid \`photon.mauthor()\` invocation:

                        {
                          include: {
                            mauthor: true,
                            ~~~~~~~
                        ?   author?: true
                          }
                        }


                        Unknown field \`mauthor\` for include statement on model Post. Available options are listed in green. Did you mean \`author\`?
                        "
                  `)
    }
  })
})
