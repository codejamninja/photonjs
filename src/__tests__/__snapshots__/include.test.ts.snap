// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`include validation allow deep include without another include 1`] = `
"query {
  findManyUser {
    id
    email
    name
    posts(first: 20) {
      id
      createdAt
      updatedAt
      published
      title
      content
    }
  }
}"
`;

exports[`include validation allow include statement with only false properties 1`] = `
"query {
  findManyPost {
    id
    createdAt
    updatedAt
    published
    title
    content
    author {
      id
      email
      name
    }
  }
}"
`;

exports[`include validation catch unknown field name 1`] = `
"query {
  mauthor {
    mauthor # INVALID_FIELD
  }
}"
`;

exports[`include validation deep include query 1`] = `
"query {
  findManyPost {
    id
    createdAt
    updatedAt
    published
    title
    content
    author {
      id
      email
      name
      posts {
        id
        createdAt
        updatedAt
        published
        title
        content
      }
    }
  }
}"
`;

exports[`include validation dont allow empty include statements 1`] = `
"query {
  findManyPost {
    include # INVALID_FIELD
  }
}"
`;

exports[`include validation handle scalar fields special 1`] = `
"query {
  id {
    id # INVALID_FIELD
  }
}"
`;
